### PR TITLE
Add support for algebraic data types

### PR DESCRIPTION
**Pull request by sigma-andex**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -105,6 +117,77 @@ sky =
...
+-- Twice to the right for CanisMaior, once to the left for the String 
+actual = preview (barlow (key :: _ "zodiac>><")) sky
+-- Just "Sirius"
+```
+
+Use `<` and `>` to zoom into the left and right cases of your product type. 
+
+```purescript
+data Zodiac
+  = Virgo { alpha :: String } { beta :: String } { gamma:: String } { delta :: String }
```
Comment (i-am-the-slime) [inline README.md:150]:
```suggestion
  = Virgo { alpha :: String } { beta :: String } { gamma :: String } { delta :: String }
```

File: README.md [hunk 2]
```diff
@@ -105,6 +117,77 @@ sky =
...
+      { virgo:
+          Virgo { alpha : "Spica"} { beta: "β Vir"} { gamma: "γ Vir B"} { delta: "δ Vir"}
+      }
+  }
+
+actual = preview (barlow (key :: _ "zodiac.virgo>>>.delta")) sky
+-- (Just "δ Vir")
+```
+
+And you can of course combine them. It is more readibly if you separate your sum lens from your product lens with a `.` dot.
```
Comment (i-am-the-slime) [inline README.md:168]:
```suggestion
And you can of course combine them. It is more readable if you separate your sum lens from your product lens with a `.` dot. 
```

File: README.md [hunk 3]
```diff
@@ -105,6 +117,77 @@ sky =
...
+
+actual = preview (barlow (key :: _ "zodiac.virgo>>>.delta")) sky
+-- (Just "δ Vir")
+```
+
+And you can of course combine them. It is more readibly if you separate your sum lens from your product lens with a `.` dot. 
+
+```purescript 
+data Zodiac
+  = Carina { alpha :: String } | Virgo { alpha :: String } { beta :: String } { gamma:: String } { delta :: String } | CanisMaior String
```
Comment (i-am-the-slime) [inline README.md:172]:
```suggestion
  = Carina { alpha :: String } | Virgo { alpha :: String } { beta :: String } { gamma :: String } { delta :: String } | CanisMaior String 
```

**Review by i-am-the-slime — APPROVED** [review]

**Comment by i-am-the-slime** [comment]
Two comments: 
 1. I'd avoid overloading operators `>` and `<` in this case
 2. I wonder if there's a way with Generic to reference the constructor name in a type-safe manner. In it's current state I'd be very careful with the ADT feature because it's not safe from reordering which shouldn't make a difference in semantics of the ADT in general

**Comment by hivemind-js** [comment]
> Two comments:
> 
>     1. I'd avoid overloading operators `>` and `<` in this case
Yes I'm also not that happy with it, but didn't see any alternative symbols.

> 
>     2. I wonder if there's a way with Generic to reference the constructor name in a type-safe manner. In it's current state I'd be very careful with the ADT feature because it's not safe from reordering which shouldn't make a difference in semantics of the ADT in general

I haven't thought about the reordering aspect. That's definitely good point. The constructor carries the symbol with it, so it should definitely possible to do something like `%Red` for an ADT `data TrafficLight = Green | Yellow | Red`. 
I was planning to make the use of `<` and `>` easier with something like `%3` in #14 , but using the symbol will probably be the better idea to make it safe w.r.t. reordering.

**Comment by hivemind-js** [comment]
Will merge it then and address the proposal in #14

**Comment by sigma-andex** [comment]
> Two comments:
> 
>     1. I'd avoid overloading operators `>` and `<` in this case

Yes I'm also not that happy with it, but didn't see any alternative symbols.

> 
>     2. I wonder if there's a way with Generic to reference the constructor name in a type-safe manner. In it's current state I'd be very careful with the ADT feature because it's not safe from reordering which shouldn't make a difference in semantics of the ADT in general

I haven't thought about the reordering aspect. That's definitely good point. The constructor carries the symbol with it, so it should definitely possible to do something like %Red for an ADT data TrafficLight = Green | Yellow | Red.
I was planning to make the use of < and > easier with something like %3 in #14 , but using the symbol will probably be the better idea to make it safe w.r.t. reordering.

**Pull request merged**